### PR TITLE
HDDS-9654. Support ranged GET request for a specified part

### PR DESCRIPTION
**Pull request by tanvipenumudy**
## What changes were proposed in this pull request?

Currently, we do not support a 'ranged' GET request for a specified part, which is useful for downloading just a part of a multipart key. Even when a part number has been specified, we still download the entire multipart key.

With this fix, we should be able to download just a part of a multipart key. The additional client logic introduced takes advantage of the existing `BlockID` to `partNumber` mapping as part of the [KeyLocation](https://github.com/apache/ozone/blob/master/hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto#L1019-L1035) proto message for filtering out the [OmKeyLocationInfo](https://github.com/apache/ozone/blob/master/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyLocationInfo.java) list.

Reference: [Amazon S3 GetObject Documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#AmazonS3-GetObject-request-querystring-PartNumber)

Before the fix:
```
aws s3api get-object --endpoint <endpoint-url>:9879 --bucket bucket --key my_1GB_file --part-number 1 outfile
{
    "AcceptRanges": "bytes",
    "LastModified": "2023-11-06T15:45:34+00:00",
    "ContentLength": 1073741824,
    "CacheControl": "no-cache",
    "ContentType": "application/octet-stream",
    "Expires": "2023-11-09T11:50:30+00:00",
    "Metadata": {}
}
```

After the fix:
```
aws s3api get-object --endpoint <endpoint-url>:9879 --bucket bucket --key my_1GB_file --part-number 1 outfile
{
    "AcceptRanges": "bytes",
    "LastModified": "2023-11-06T15:45:34+00:00",
    "ContentLength": 500000000,
    "CacheControl": "no-cache",
    "ContentType": "application/octet-stream",
    "Expires": "2023-11-09T11:46:20+00:00",
    "Metadata": {}
}
```

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9654

## How was this patch tested?

- Added integration tests
- Green CI/CD checks: https://github.com/tanvipenumudy/ozone/actions/runs/6807544806/

### CODE REVIEW

**Comment by kerneltime** [comment]
@SaketaChalamchala @xBis7 @vtutrinov can you please take a look as well?

**Comment by vtutrinov** [comment]
LGTM
@tanvipenumudy thank you for adding the feature.

**Comment by kerneltime** [comment]
Thank you @tanvipenumudy for this change and @vtutrinov for the review. @tanvipenumudy can you please extend the robot test to read a part and validate the size read?

**Review by vtutrinov — APPROVED** [review]

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 1]
```diff
@@ -1702,7 +1687,19 @@ public OzoneKeyDetails getS3KeyDetails(String bucketName, String keyName,
...
+
+  private long getTotalBytesRead(OmKeyInfo keyInfo) throws IOException {
+    OzoneInputStream keyContent = getInputStreamWithRetryFunction(keyInfo);
+    byte[] buffer = new byte[4096];
+    int bytesRead;
+    long totalBytesRead = 0;
+    while ((bytesRead = keyContent.read(buffer)) != -1) {
+      totalBytesRead += bytesRead;
+    }
+    return totalBytesRead;
```
Comment (adoroszlai) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1702]: Does this mean client reads data for the specified part just to calculate size for `ContentLength`?

Comment (tanvipenumudy) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1702]: That's right, I was not able to see any existing way in which the OmKeyInfo#dataSize could dynamically get updated based on the updated `OmKeyLocationInfo` list.

I am open to suggestions if there is a better way to handle this!

Comment (tanvipenumudy) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1702]: As discussed offline with @kerneltime, I think we can have the block lengths summed up for a given `partNumber` instead - I will update the changes accordingly, thanks!

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @tanvipenumudy for the patch, @kerneltime, @vtutrinov for the review.

**Pull request merged**